src/Fmt_ast.ml

```diff
@@ -3710,16 +3740,47 @@ and fmt_structure c ctx itms =
           | Pmod_ident _ -> true
           | _ -> false
         in
-        let is_simple itm =
-          match itm.pstr_desc with
-          | Pstr_include {pincl_mod= me} | Pstr_module {pmb_expr= me} ->
-              is_simple_mod me
-          | Pstr_open _ -> true
-          | _ -> false
+        let is_simple (itm, c) =
+          match c.conf.structure_item_grouping with
+          | `Compact ->
+              Location.width itm.pstr_loc <= c.conf.margin
+              && Location.is_single_line itm.pstr_loc
+          | `Sparse -> (
+            match itm.pstr_desc with
+            | Pstr_include {pincl_mod= me} | Pstr_module {pmb_expr= me} ->
+                is_simple_mod me
+            | Pstr_open _ -> true
+            | _ -> false )
+        in
+        let allow_adjacent (itmI, cI) (itmJ, cJ) =
```

Comment:
To make it easier to keep structures and signatures in sync, it would help to refactor this and the analogous function for signatures above to one that operates on Ast.t. This probably should go near the beginning of this module, after the sugar_ functions.